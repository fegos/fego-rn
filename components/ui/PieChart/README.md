---
title: PieChart
subTitle: 饼图
---

### 动画扇形图
### 注意事项
```warning
使用的页面必须有`state`初始化，可以为空，用来触发动画的绘制。
若从外部传入参数，则建议使用

componentDidUpdate(prevProps) {
  const nextProps=this.props;
	if (nextProps) {
		this.setState({
			percent: nextProps.percent,
		})
		setTimeout(() => {
			this.setState({
				percent: nextProps.percent,
			})
		}, 0);
	}
}

来设定传入参数，触发动画。
```

### example
```
<PieChart
	percentArray={[0.4, 0.6]}
	colorArray={['#4d84eb', '#fca63e']}
	outerRadius={40}
	innerRadius={25}
	duration={1000}
/>
```
### 属性
```
PieChart.propTypes = {
	percentArray: PropTypes.array.isRequired,
	colorArray: PropTypes.array.isRequired,
	innerRadius: PropTypes.number,
	outerRadius: PropTypes.number.isRequired,
	duration: PropTypes.number,
};
PieChart.defaultProps = {
	innerRadius: 0,
	duration: 1500,
}
```
### 属性定义

| 名称 | 类型 | 描述 |
|------|------|-------------|
| `percentArray` | `array` | 扇形各段的百分比，需自行计算，1为整个圆形图案 ，必填|
| `colorArray` | `array` | 扇形各段的颜色，按顺序，与percentArray数量须相同 ，必填|
| `innerRadius` | `number` | 扇形内环半径，默认为0（圆形），扇形效果需设置，需小于外环半径|
| `outerRadius` | `number` | 扇形外环半径，必填|
| `duration` | `number` | 每段动画时长，默认1500毫秒 |
| `configArray` | `array` | 其他配置，eg: `[,{stroke:'red',strokeWidth:1,strokeDash:[2,5]},,{stroke:'black',strokeWidth:1,strokeDash:[2,5]}]` |
