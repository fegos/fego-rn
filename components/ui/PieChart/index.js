import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { ART, Animated } from 'react-native';
import Wedge from './Wedge';

const { Surface, Group } = ART;

const AnimatedWedge = Animated.createAnimatedComponent(Wedge);
const circumference = 360;
export default class PieChart extends Component {
  static propTypes = {
    // 饼图数据
    percentArray: PropTypes.arrayOf(PropTypes.any).isRequired,
    // 饼图颜色
    colorArray: PropTypes.arrayOf(PropTypes.any).isRequired,
    // 饼图内直径
    innerRadius: PropTypes.number,
    // 饼图外直径
    outerRadius: PropTypes.number.isRequired,
    // 动画执行方式
    animationType: PropTypes.oneOf(['sequence', 'synchron']),
    // 开始动画的角度
    rotation: PropTypes.number,
    // 动画执行方向，是否为顺时针
    isClockwise: PropTypes.bool,
    // 动画执行时间
    duration: PropTypes.number,
    // 配置, eg: [,{stroke:'red',strokeWidth:1,strokeDash:[2,5]},,{stroke:'black',strokeWidth:1,strokeDash:[2,5]}]
    configArray: PropTypes.arrayOf(PropTypes.any),
    // 动画结束时的回调函数
    animationEndCallBack: PropTypes.func,
  }

  static defaultProps = {
    // 饼图内直径
    innerRadius: 0,
    // 动画执行时间
    duration: 1500,
    // 开始动画的角度
    rotation: 0,
    // 配置, eg: [,{stroke:'red',strokeWidth:1,strokeDash:[2,5]},,{stroke:'black',strokeWidth:1,strokeDash:[2,5]}]
    configArray: [],
    // 顺序执行
    animationType: 'sequence',
    // 顺时针显示
    isClockwise: true,
    animationEndCallBack: null,
  }

  constructor(props) {
    super(props);

    this.animationArray = [];
    this.endAngleArray = [];

    this.state = {
      wedgeAngles: [],
    };
  }

  componentDidMount() {
    this._runPieChart();
  }

  componentDidUpdate(prevProps, prevState, snapshot) {
    super.componentDidUpdate(prevProps);
    if (snapshot) {
      this._runPieChart();
    }
  }

  getSnapshotBeforeUpdate(prevProps) {
    const nextProps = this.props;
    if (nextProps && prevProps) {
      if ((nextProps.percentArray.toString() === prevProps.percentArray.toString()) &&
        (nextProps.colorArray.toString() === prevProps.colorArray.toString())) {
        return null;
      }
    }
    return nextProps;
  }

  _runPieChart = () => {
    // 初始化动画对象
    this.animationArray = [];
    for (let index = 0; index < this.props.percentArray.length; index++) {
      this.animationArray.push(new Animated.Value(0));
    }
    this._handleData();
    this._animations();
  }

  _sequenceAnimation = () => {
    const animatedArray = [];
    for (let index = 0; index < this.props.percentArray.length; index++) {
      animatedArray.push(Animated.timing(this.animationArray[index], {
        duration: this.props.duration,
        toValue: circumference,
      }));
    }
    Animated.sequence(animatedArray).start(this.props.animationEndCallBack);
  }
  _syncAnimation = () => {
    const animatedArray = [];
    for (let index = 0; index < this.props.percentArray.length; index++) {
      animatedArray.push(Animated.timing(this.animationArray[index], {
        duration: this.props.duration,
        toValue: circumference,
      }));
    }
    Animated.parallel(animatedArray).start(this.props.animationEndCallBack);
  }

  _animations = () => {
    this.props.animationType === 'sequence' ? this._sequenceAnimation() : this._syncAnimation();
  }

  _handleData = () => {
    const wedgeAngles = [];
    const endAngleArray = [];

    // 处理百分比，得到每个部分的结束位置
    for (let index = 0; index < this.props.percentArray.length; index++) {
      let sum = 0;
      for (let index2 = 0; index2 <= index; index2++) {
        sum += this.props.percentArray[index2];
      }
      this.props.isClockwise ? endAngleArray.push(sum) : endAngleArray.push(1 - sum);
    }
    this.endAngleArray = endAngleArray;

    // 添加动画对象数组
    for (let index = 0; index < this.props.percentArray.length; index++) {
      // 起始角度
      const startAngle = index === 0 ? 0 : this.endAngleArray[index - 1] * circumference;
      // 结束角度

      const endAngle = this.props.isClockwise ?
        startAngle + (this.props.percentArray[index] * circumference) :
        startAngle - (this.props.percentArray[index] * circumference);

      wedgeAngles.push(this.animationArray[index].interpolate({
        inputRange: [0, circumference],
        outputRange: [startAngle, endAngle],
        extrapolate: 'clamp',
      }));
    }
    this.setState({
      wedgeAngles,
    });
  }

  render() {
    return (
      <Surface rotation={-90} width={this.props.outerRadius * 2} height={this.props.outerRadius * 2}>
        <Group rotation={this.props.rotation} originX={this.props.outerRadius} originY={this.props.outerRadius}>
          {this.state.wedgeAngles.map((data, index) => {
            let rstroke;
            let rstrokeWidth;
            let rstrokeDash;
            if (this.props.configArray.length > 0 && this.props.configArray[index]) {
              rstroke = this.props.configArray[index].stroke;
              rstrokeWidth = this.props.configArray[index].strokeWidth;
              rstrokeDash = this.props.configArray[index].strokeDash;
            }
            const startAngle = index === 0 ? index : this.endAngleArray[index - 1] * circumference;
            const endAngle = this.state.wedgeAngles[index];
            const key = index;
            return (
              <AnimatedWedge
                key={key}
                outerRadius={this.props.outerRadius}
                innerRadius={this.props.innerRadius}
                startAngle={this.props.isClockwise ? startAngle : endAngle}
                stroke={rstroke}
                strokeWidth={rstrokeWidth}
                strokeDash={rstrokeDash}
                endAngle={this.props.isClockwise ? endAngle : startAngle}
                fill={this.props.colorArray[index]}
              />
            );
          })}
        </Group>
      </Surface>
    );
  }
}

PieChart.Wedge = Wedge;
