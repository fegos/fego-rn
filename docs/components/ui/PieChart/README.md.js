webpackJsonp([18],{484:function(n,t){n.exports={content:["article",["h3","\u52a8\u753b\u6247\u5f62\u56fe"],["h3","\u6ce8\u610f\u4e8b\u9879"],["pre",{lang:"warning",highlighted:'\u4f7f\u7528\u7684\u9875\u9762\u5fc5\u987b\u6709`state`\u521d\u59cb\u5316\uff0c\u53ef\u4ee5\u4e3a\u7a7a\uff0c\u7528\u6765\u89e6\u53d1\u52a8\u753b\u7684\u7ed8\u5236\u3002\n\u82e5\u4ece\u5916\u90e8\u4f20\u5165\u53c2\u6570\uff0c\u5219\u5efa\u8bae\u4f7f\u7528\n\n<span class="token function">componentWillReceiveProps</span><span class="token punctuation">(</span>nextProps<span class="token punctuation">)</span> {\n\t<span class="token keyword">if</span> <span class="token punctuation">(</span>nextProps<span class="token punctuation">)</span> {\n\t\tthis<span class="token punctuation">.</span><span class="token function">setState</span><span class="token punctuation">(</span>{\n\t\t\tpercent<span class="token punctuation">:</span> nextProps<span class="token punctuation">.</span>percent<span class="token punctuation">,</span>\n\t\t}<span class="token punctuation">)</span>\n\t\t<span class="token function">setTimeout</span><span class="token punctuation">(</span><span class="token punctuation">(</span><span class="token punctuation">)</span> <span class="token operator">=</span><span class="token operator">></span> {\n\t\t\tthis<span class="token punctuation">.</span><span class="token function">setState</span><span class="token punctuation">(</span>{\n\t\t\t\tpercent<span class="token punctuation">:</span> nextProps<span class="token punctuation">.</span>percent<span class="token punctuation">,</span>\n\t\t\t}<span class="token punctuation">)</span>\n\t\t}<span class="token punctuation">,</span> <span class="token number">0</span><span class="token punctuation">)</span><span class="token comment" spellcheck="true">;</span>\n\n\t}\n}\n\n\u6765\u8bbe\u5b9a\u4f20\u5165\u53c2\u6570\uff0c\u89e6\u53d1\u52a8\u753b\u3002'},["code","\u4f7f\u7528\u7684\u9875\u9762\u5fc5\u987b\u6709`state`\u521d\u59cb\u5316\uff0c\u53ef\u4ee5\u4e3a\u7a7a\uff0c\u7528\u6765\u89e6\u53d1\u52a8\u753b\u7684\u7ed8\u5236\u3002\n\u82e5\u4ece\u5916\u90e8\u4f20\u5165\u53c2\u6570\uff0c\u5219\u5efa\u8bae\u4f7f\u7528\n\ncomponentWillReceiveProps(nextProps) {\n\tif (nextProps) {\n\t\tthis.setState({\n\t\t\tpercent: nextProps.percent,\n\t\t})\n\t\tsetTimeout(() => {\n\t\t\tthis.setState({\n\t\t\t\tpercent: nextProps.percent,\n\t\t\t})\n\t\t}, 0);\n\n\t}\n}\n\n\u6765\u8bbe\u5b9a\u4f20\u5165\u53c2\u6570\uff0c\u89e6\u53d1\u52a8\u753b\u3002"]],["h3","example"],["pre",{lang:null,highlighted:'<span class="token operator">&lt;</span>PieChart\n\tpercentArray<span class="token operator">=</span>{<span class="token punctuation">[</span><span class="token number">0.4</span><span class="token punctuation">,</span> <span class="token number">0.6</span><span class="token punctuation">]</span>}\n\tcolorArray<span class="token operator">=</span>{<span class="token punctuation">[</span><span class="token string">\'#4d84eb\'</span><span class="token punctuation">,</span> <span class="token string">\'#fca63e\'</span><span class="token punctuation">]</span>}\n\touterRadius<span class="token operator">=</span>{<span class="token number">40</span>}\n\tinnerRadius<span class="token operator">=</span>{<span class="token number">25</span>}\n\tduration<span class="token operator">=</span>{<span class="token number">1000</span>}\n<span class="token operator">/</span><span class="token operator">></span>'},["code","<PieChart\n\tpercentArray={[0.4, 0.6]}\n\tcolorArray={['#4d84eb', '#fca63e']}\n\touterRadius={40}\n\tinnerRadius={25}\n\tduration={1000}\n/>"]],["h3","\u5c5e\u6027"],["pre",{lang:null,highlighted:'PieChart<span class="token punctuation">.</span>propTypes <span class="token operator">=</span> {\n\tpercentArray<span class="token punctuation">:</span> PropTypes<span class="token punctuation">.</span>array<span class="token punctuation">.</span>isRequired<span class="token punctuation">,</span>\n\tcolorArray<span class="token punctuation">:</span> PropTypes<span class="token punctuation">.</span>array<span class="token punctuation">.</span>isRequired<span class="token punctuation">,</span>\n\tinnerRadius<span class="token punctuation">:</span> PropTypes<span class="token punctuation">.</span>number<span class="token punctuation">,</span>\n\touterRadius<span class="token punctuation">:</span> PropTypes<span class="token punctuation">.</span>number<span class="token punctuation">.</span>isRequired<span class="token punctuation">,</span>\n\tduration<span class="token punctuation">:</span> PropTypes<span class="token punctuation">.</span>number<span class="token punctuation">,</span>\n}<span class="token comment" spellcheck="true">;</span>\nPieChart<span class="token punctuation">.</span>defaultProps <span class="token operator">=</span> {\n\tinnerRadius<span class="token punctuation">:</span> <span class="token number">0</span><span class="token punctuation">,</span>\n\tduration<span class="token punctuation">:</span> <span class="token number">1500</span><span class="token punctuation">,</span>\n}'},["code","PieChart.propTypes = {\n\tpercentArray: PropTypes.array.isRequired,\n\tcolorArray: PropTypes.array.isRequired,\n\tinnerRadius: PropTypes.number,\n\touterRadius: PropTypes.number.isRequired,\n\tduration: PropTypes.number,\n};\nPieChart.defaultProps = {\n\tinnerRadius: 0,\n\tduration: 1500,\n}"]],["h3","\u5c5e\u6027\u5b9a\u4e49"],["table",["thead",["tr",["th","\u540d\u79f0"],["th","\u7c7b\u578b"],["th","\u63cf\u8ff0"]]],["tbody",["tr",["td",["code","percentArray"]],["td",["code","array"]],["td","\u6247\u5f62\u5404\u6bb5\u7684\u767e\u5206\u6bd4\uff0c\u9700\u81ea\u884c\u8ba1\u7b97\uff0c1\u4e3a\u6574\u4e2a\u5706\u5f62\u56fe\u6848 \uff0c\u5fc5\u586b"]],["tr",["td",["code","colorArray"]],["td",["code","array"]],["td","\u6247\u5f62\u5404\u6bb5\u7684\u989c\u8272\uff0c\u6309\u987a\u5e8f\uff0c\u4e0epercentArray\u6570\u91cf\u987b\u76f8\u540c \uff0c\u5fc5\u586b"]],["tr",["td",["code","innerRadius"]],["td",["code","number"]],["td","\u6247\u5f62\u5185\u73af\u534a\u5f84\uff0c\u9ed8\u8ba4\u4e3a0\uff08\u5706\u5f62\uff09\uff0c\u6247\u5f62\u6548\u679c\u9700\u8bbe\u7f6e\uff0c\u9700\u5c0f\u4e8e\u5916\u73af\u534a\u5f84"]],["tr",["td",["code","outerRadius"]],["td",["code","number"]],["td","\u6247\u5f62\u5916\u73af\u534a\u5f84\uff0c\u5fc5\u586b"]],["tr",["td",["code","duration"]],["td",["code","number"]],["td","\u6bcf\u6bb5\u52a8\u753b\u65f6\u957f\uff0c\u9ed8\u8ba41500\u6beb\u79d2"]],["tr",["td",["code","configArray"]],["td",["code","array"]],["td","\u5176\u4ed6\u914d\u7f6e\uff0ceg: ",["code","[,{stroke:'red',strokeWidth:1,strokeDash:[2,5]},,{stroke:'black',strokeWidth:1,strokeDash:[2,5]}]"]]]]]],meta:{title:"PieChart",subTitle:"\u997c\u56fe",filename:"components/ui/PieChart/README.md"},toc:["ul",["li",["a",{className:"bisheng-toc-h3",href:"#\u52a8\u753b\u6247\u5f62\u56fe",title:"\u52a8\u753b\u6247\u5f62\u56fe"},"\u52a8\u753b\u6247\u5f62\u56fe"]],["li",["a",{className:"bisheng-toc-h3",href:"#\u6ce8\u610f\u4e8b\u9879",title:"\u6ce8\u610f\u4e8b\u9879"},"\u6ce8\u610f\u4e8b\u9879"]],["li",["a",{className:"bisheng-toc-h3",href:"#example",title:"example"},"example"]],["li",["a",{className:"bisheng-toc-h3",href:"#\u5c5e\u6027",title:"\u5c5e\u6027"},"\u5c5e\u6027"]],["li",["a",{className:"bisheng-toc-h3",href:"#\u5c5e\u6027\u5b9a\u4e49",title:"\u5c5e\u6027\u5b9a\u4e49"},"\u5c5e\u6027\u5b9a\u4e49"]]]}}});