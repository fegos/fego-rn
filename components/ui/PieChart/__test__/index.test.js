import React from 'react';
import Renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import PieChart from '../index';

jest.useFakeTimers();
describe('PieChart Tests', () => {
  let Comp;
  beforeEach(() => {
    Comp = (
      <PieChart
        percentArray={[0.2, 0.1, 0.4, 0.3]}
        colorArray={['red', 'yellow', 'blue', 'green']}
        outerRadius={40}
        innerRadius={25}
        duration={1000}
      />
    );
  });

  it('renders correctly', () => {
    const tree = Renderer.create(Comp).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('other', () => {
    const wrapper = shallow(Comp);
    expect(wrapper.instance().props.innerRadius).toBe(25);
  });
});
