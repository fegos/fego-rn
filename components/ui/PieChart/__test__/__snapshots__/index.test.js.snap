// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PieChart Tests renders correctly 1`] = `
<ARTSurfaceView
  style={
    Array [
      undefined,
      Object {
        "height": 80,
        "width": 80,
      },
    ]
  }
>
  <ARTGroup
    opacity={1}
    transform={
      Array [
        1,
        0,
        0,
        1,
        0,
        0,
      ]
    }
  >
    <ARTShape
      d={Array []}
      fill={null}
      opacity={1}
      stroke={null}
      strokeCap={1}
      strokeDash={null}
      strokeJoin={1}
      strokeWidth={1}
      transform={
        Array [
          1,
          0,
          0,
          1,
          0,
          0,
        ]
      }
    />
    <ARTShape
      d={Array []}
      fill={null}
      opacity={1}
      stroke={null}
      strokeCap={1}
      strokeDash={null}
      strokeJoin={1}
      strokeWidth={1}
      transform={
        Array [
          1,
          0,
          0,
          1,
          0,
          0,
        ]
      }
    />
    <ARTShape
      d={Array []}
      fill={null}
      opacity={1}
      stroke={null}
      strokeCap={1}
      strokeDash={null}
      strokeJoin={1}
      strokeWidth={1}
      transform={
        Array [
          1,
          0,
          0,
          1,
          0,
          0,
        ]
      }
    />
    <ARTShape
      d={Array []}
      fill={null}
      opacity={1}
      stroke={null}
      strokeCap={1}
      strokeDash={null}
      strokeJoin={1}
      strokeWidth={1}
      transform={
        Array [
          1,
          0,
          0,
          1,
          0,
          0,
        ]
      }
    />
  </ARTGroup>
</ARTSurfaceView>
`;
